src/udev/udev-node.c

```diff
@@ -188,28 +201,116 @@ static int link_find_prioritized(sd_device *dev, bool add, const char *stackdir,
         return 0;
 }
 
+static int init_link_semaphores(const char *path) {
+        key_t key = ftok(path, 0);
+        int semid;
+
+        semid = semget(key, N_SEMAPHORES, 0600|IPC_CREAT|IPC_EXCL);
+        if (semid != -1) {
+                unsigned short val[N_SEMAPHORES];
+                int i;
+                struct sembuf dummy_op[]  = {
+                        { .sem_op = -1, },
+                        { .sem_op = 1, },
+                };
+
+                for (i = 0; i < N_SEMAPHORES; i++)
+                        val[i] = 1;
+                if (semctl(semid, 0, SETALL, val) == -1)
+                        return log_error_errno(-errno, "Failed to initialize semaphores: %m");
+
+                /* Dummy semop to set sem_otime */
+                if (semop(semid, dummy_op, (sizeof(dummy_op)/sizeof(*dummy_op))) == -1)
+                        return log_error_errno(-errno, "Failed to set sem_otime: %m");
+                return semid;
+        } else {
+                const unsigned int RETRIES = 10, SLEEP_US = 10000;
+                unsigned int i;
+
+                semid = semget(key, N_SEMAPHORES, 0);
+                if (semid == -1)
+                        return log_error_errno(-errno, "Failed to get semaphore set: %m");
+
+                for (i = 0; i < RETRIES; i++) {
+                        struct semid_ds ds;
+
+                        /* Wait for initialization to finish */
+                        if (semctl(semid, 0, IPC_STAT, &ds) != -1 &&
+                            ds.sem_otime != 0)
+                                return semid;
+                        usleep(SLEEP_US);
+                }
```

Comment:
busy looping? urks!

Reply:
I agree that it's not beautiful, but there's no event we can listen for. This loop is only executed if there's a race between two processes __trying to initialize__ the semaphore set at the same time, that is, at most once during a system's uptime. AFAIK, this is how SysV semaphore initalization is supposed to be done. (_FTR, as long as usleep() is called, I wouldn't call it busy-looping_).

I guess what I could do instead is use some other locking mechanism, e.g. a file lock, just for the purpose of running semget(). That way we'd ensure that the semaphore set would either not exist or be fully initialized when semget() is installed. From what I saw in earlier tests though, even that file lock, which must be taken once by every udev worker, might cause a scaling problem.

But it seems that you'd reject any SysV IPC-based  approach anyway, so does it make sense for me to bother?

Reply:
yeah, sysvipc is not the way to go, sorry. the usleep() thing is just one issue among many with sysvipc...

But again, I don't see why the renameat() loop proposal I gave wouldn't work. What's essential is that the determination which symlink to use is enclosed with the readlinkat() before and the renameat() after and repeated until the result of the latter matches what we expect based on the results of the former. can you provide me with a runthrough ordering where this wouldn't work correctly?